controlpanel/api/tools.py

```diff
@@ -0,0 +1,237 @@
+from collections import UserDict
+import secrets
+
+from django.conf import settings
+from django.utils.functional import cached_property
+from rest_framework.exceptions import APIException
+
+from controlpanel.api.helm import Helm
+from controlpanel.kubeapi.views import client
+
+
+IDLED = "mojanalytics.xyz/idled"
+
+SUPPORTED_TOOL_NAMES = list(settings.TOOLS.keys())
+
+
+class DuplicateTool(Exception):
+    pass
+
+
+class ToolNotDeployed(APIException):
+    status_code = 400
+    default_detail = f"Specified tool not deployed"
+    default_code = "tool_not_deployed"
+
+
+class UnsupportedToolException(APIException):
+    status_code = 400
+    supported = ", ".join(SUPPORTED_TOOL_NAMES)
+    default_detail = f"Unsupported tool, must be one of: {supported}"
+    default_code = "unsupported_tool"
+
+
+class ToolRepository(UserDict):
+
+    def __getitem__(self, key):
+        if key not in self.tools:
+            raise UnsupportedToolException(key)
+
+        if key not in self.tools:
+            raise UnsupportedToolException(key)
+
+        return super().__getitem__(key)
+
+    def __setitem__(self, key, value):
+        if key not in SUPPORTED_TOOL_NAMES:
+            raise UnsupportedToolException(key)
+
+        if key in self.tools:
+            raise DuplicateTool(key)
+
+        super().__setitem__(key, value)
```

Comment:
#### 1e. Refactor Tools code
> Lines 34 - 52

I updated the ToolRepository class to disallow adding unsupported tool names or duplicates

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-aad7b41260663fb400ff942a02fd3d14R55)